platforms/jvm/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoTaskExtension.java

```diff
@@ -125,7 +126,10 @@ public File getDestinationFile() {
      * @since 4.0
      */
     public void setDestinationFile(Provider<File> destinationFile) {
-        this.destinationFile.fileProvider(destinationFile);
+        // TODO: This is a workaround for behavior in AGP.
+        // see https://github.com/gradle/gradle/issues/33389
+        // This can be removed once we've fixed RegularFileProperty.fileProvider(...) to work properly 
+        this.destinationFile.fileProvider(destinationFile.flatMap(Providers::of));
```

Comment:
I'd love to have a test for this, but it's complicated to setup in a reasonable way:
1. We need to have a task that has an output file
2. We need to wire that output file to this property
3. We need the same task's action to pull on _this_ property 

Alternatively, maybe I can add something to a smoke test to exercise this by enabling code coverage.  I haven't tried that yet.